docs/development/virtual_environments.rst

```diff
@@ -315,3 +328,116 @@ By default, direct SSH access is not enabled in the prod environment. You will n
 in over Tor after initial provisioning or set ``enable_ssh_over_tor`` to "false"
 during ``./securedrop-admin tailsconfig``. See :ref:`ssh_over_tor` or :ref:`ssh_over_local`
 for more info.
+
+.. _molecule_upgrade:
+
+Molecule: upgrade testing
+-------------------------
+The Molecule upgrade scenario sets up a predefined staging Securedrop virtual 
+environment using Vagrant boxes built with the latest application release.
+It also creates a virtualized APT repository, and modifies 
+the SecureDrop environment to use this APT repository instead of the FPF main 
+repo at https://apt.freedom.press/. 
+
+You can use this scenario to test the upgrade process, using using either 
+locally-built .debs or packages from the FPF test repo at 
+https://apt-test.freedom.press/. Both options are described below.
+  
+.. note:: The upgrade scenario uses QEMU/KVM via Vagrant's libvirt provider, in 
+   place of the  default Virtualbox provider. If you haven't already done so, 
+   you'll need to set up the libvirt provider before proceeding. For 
+   more information, see :ref:`libvirt_provider`.
+
+Upgrade testing using locally-built packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, build the app code packages and create the environment:
+
+.. code:: sh
+
+ make build-debs
+ molecule converge -s upgrade
+
+The playbook will return the source interface Onion address. You can use this to
+check the application version displayed in the source interface footer. 
+Alternatively, you can log into the application server VM and check the deployed 
+package version directly:
+
+.. code:: sh
+
+   molecule login -s upgrade -h app-staging
+
+From the application server:
```

Comment:
Nit: terminology usage such as "application server" should follow [formatting recommendations](https://docs.securedrop.org/en/latest/development/documentation_guidelines.html#terminology); here, that means writing _Application Server_. Those guidelines are intended primarily to be used when referring to the actual SecureDrop servers, rather than their VM surrogates, but still strikes me as a reasonable improvement here for clarity.

Reply:
:heavy_check_mark: